### PR TITLE
Deprecate usage of `accountId` in new model classes

### PR DESCRIPTION
**Pull request by benwerner01**
## 🌟 What is the purpose of this PR?

This PR replaces the usage of `accountId` with `ownedById` in the new model classes.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1202805690238892/1203063464592800/f) _(internal)_

## 🔍 What does this change?

See the commits and PR comments.

## 📜 Does this require a change to the docs?

No.

## 🛡 What tests cover this?

- the existing integration tests

## ❓ How to test this?

Verify the integration tests succeed in CI, or by running them locally.

### CODE REVIEW

File: packages/hash/api/src/graphql/resolvers/entityType/createEntityType.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ export const deprecatedCreateEntityType: ResolverFn<
   dataSources.db.transaction(async (client) => {
     const entityType = await EntityType.create(client, {
       accountId,
-      createdByAccountId: user.accountId,
+      createdByAccountId: user.entityId,
```
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/entityType/createEntityType.ts:21]: The reason I've had to touch the legacy GQL resolvers in this PR is because we are attaching the updated `UserModel` class to the GQL context, so changes to its type propagate into these files. These uses of `user.accountId` were incorrect to begin with, so have replaced them with `user.entityId` which is closer to what we will end up with.

File: packages/hash/api/src/graphql/resolvers/middlewares/canAccessAccount.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { ResolverMiddleware } from "./middlewareTypes";
 export const canAccessAccount: ResolverMiddleware<
```
Comment (benwerner01) [inline packages/hash/api/src/graphql/resolvers/middlewares/canAccessAccount.ts:11]: We are not yet using this middleware function with any of the re-implemented GQL mutations/queries, but have updated it here in case we do want to make use of it.

File: packages/hash/api/src/model/knowledge/entity.model.ts [hunk 3]
```diff
@@ -284,13 +283,11 @@ export default class {
   /**
    * Get the latest version of an entity by its entity ID.
    *
-   * @param params.accountId the accountId of the account requesting the entity
    * @param params.versionedUri the unique versioned URI for an entity.
    */
   static async getLatest(
     graphApi: GraphApi,
     params: {
-      accountId?: string;
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/entity.model.ts:293]: I've removed `accountId` from methods where it is currently not being used. Once we start sharding on `ownedById`, the field should be re-introduced. The type errors will remind us to do this, so I didn't think it was worth adding a todo

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 4]
```diff
@@ -121,14 +121,7 @@ export default class {
...
       {
-        /**
-         * @todo figure out what account ID we use here
-         *   Directly related to
-         *   https://app.asana.com/0/1202805690238892/1202883599104674/f
-         *   And may require consideration for
-         *   https://app.asana.com/0/1202805690238892/1202890446280569/f
-         */
-        ownedById: createdById,
+        ownedById,
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/link.model.ts:124]: @thehabbos007 not sure what the purpose of this comment was. The `ownedById` here is just the owner of the link that's being created, just like the `ownedById` for the create entity method is the owner of the entity that's being created.

If there's something we need to figure out here, could you explain so we can update this comment to reflect that?

Apologies if I've misunderstood it's purpose all together.

File: packages/hash/api/src/model/knowledge/link.model.ts [hunk 5]
```diff
@@ -354,7 +356,7 @@ export default class {
   /**
    * Remove the link.
    *
-   * @param removedbyId - the account removing the link
+   * @param removedbyId - the id of the user removing the link
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/link.model.ts:359]: I didn't remove `removedById` because it's already a required provenance field on `graphApi.removeLink`. This propagates into some related methods, which is why we have `updatedById` for the `updateWithoutUpdatingSiblings ` method for instance.

File: packages/hash/api/src/model/knowledge/orgMembership.model.ts [hunk 6]
```diff
@@ -22,10 +22,24 @@ export type OrgMembershipModelCreateParams = Omit<
...
+      entity.entityTypeModel.schema.$id !==
+      WORKSPACE_TYPES.entityType.orgMembership.schema.$id
+    ) {
+      throw new Error(
+        `Entity with id ${entity.entityId} is not a workspace org membership`,
+      );
+    }
+
+    return new OrgMembershipModel(entity);
+  }
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/orgMembership.model.ts:36]: This method was missing, so added it for consistency as a drive-by

File: packages/hash/api/src/model/knowledge/user.model.ts [hunk 7]
```diff
@@ -30,57 +29,6 @@ type UserModelCreateParams = Omit<
  * @class {@link UserModel}
  */
 export default class extends EntityModel {
-  /**
-   * @todo: This method and `getUserByEntityId` is confusing. Should be fixed as part of:
-   *  https://app.asana.com/0/1200211978612931/1202937382769276/f
-   */
-  static async getUserByAccountId(
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/user.model.ts:37]: This should have been removed in a prior PR, so have removed it here as a drive-by.

File: packages/hash/api/src/model/knowledge/user.model.ts [hunk 8]
```diff
@@ -30,57 +29,6 @@ type UserModelCreateParams = Omit<
...
-
-    return matchingUser ?? null;
-  }
-
-  /**
-   * Get a workspace user entity by their entityId.
-   *
-   * @param params.entityId - the entityId of the user
-   */
-  static async getUserByEntityId(
```
Comment (benwerner01) [inline packages/hash/api/src/model/knowledge/user.model.ts:66]: This method's implementation is identical to `getUserById` (except for its return-type, which was incorrect anyway), so have removed it.

File: packages/hash/api/src/model/ontology/data-type.model.ts [hunk 9]
```diff
@@ -59,35 +59,23 @@ export default class {
...
-          )?.getShortname();
-
-    if (namespace == null) {
-      throw new Error(
-        `failed to get namespace for account: ${params.accountId}`,
-      );
-    }
+    const namespace = await getNamespaceOfAccountOwner(graphApi, {
+      ownerId: params.ownedById,
+    });
```
Comment (benwerner01) [inline packages/hash/api/src/model/ontology/data-type.model.ts:78]: This code was duplicated across all the ontology model classes, so have encapsulated it in `getNamespaceOfAccountOwner`

File: packages/hash/api/src/model/ontology/util.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  const namespace =
+    params.ownerId === workspaceAccountId
+      ? WORKSPACE_ACCOUNT_SHORTNAME
+      : (
+          (await UserModel.getUserById(graphApi, {
+            entityId: params.ownerId,
+          }).catch(() => undefined)) ??
+          (await OrgModel.getOrgById(graphApi, {
+            entityId: params.ownerId,
+          }).catch(() => undefined))
```
Comment (benwerner01) [inline packages/hash/api/src/model/ontology/util.ts:25]: Previously we weren't accounting for types being created in org accounts, so have accounted for that here as a drive-by

File: packages/hash/api/src/model/knowledge/link.model.ts (hunk 4 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/knowledge/link.model.ts:124]: This comment seems to have found its way into another location that it was originally. I don't remember this being here :sweat_smile: 

I think it was a previous usage not having a clear accountId to use for the request. Don't think it's that relevant anymore as we do have a clear owner of the link in this situation

File: packages/hash/api/src/model/ontology/data-type.model.ts [hunk 11]
```diff
@@ -162,24 +149,29 @@ export default class {
...
     },
   ): Promise<DataTypeModel> {
-    const { accountId, schema } = params;
+    const { schema } = params;
 
     const updateArguments: UpdateDataTypeRequest = {
-      accountId,
+      /**
+       * @todo: let caller update who owns the type, or create new method dedicated to changing the owner of the type
+       * @see https://app.asana.com/0/1202805690238892/1203063463721791/f
```
Comment (thehabbos007) [inline packages/hash/api/src/model/ontology/data-type.model.ts:169]: This linked task doesn't seem to be the intended task (it's the same as the todo below).

File: packages/hash/api/src/model/ontology/data-type.model.ts [hunk 12]
```diff
@@ -162,24 +149,29 @@ export default class {
...
+       *
+       * @todo: replace uses of `accountId` with `ownedById` in the Graph API
+       * @see https://app.asana.com/0/1202805690238892/1203063463721791/f
```
Comment (thehabbos007) [inline packages/hash/api/src/model/ontology/data-type.model.ts:172]: Should this asana task be addressed more towards the graph API than the model classes?

File: packages/hash/api/src/model/ontology/util.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
+import { WORKSPACE_ACCOUNT_SHORTNAME } from "@hashintel/hash-backend-utils/system";
+import { OrgModel, UserModel } from "..";
+import { GraphApi } from "../../graph";
+import { workspaceAccountId } from "../util";
+
+/**
+ * Get the namespace of an account owner by its id
+ *
+ * @param params.ownerId - the id of the owner
```
Comment (thehabbos007) [inline packages/hash/api/src/model/ontology/util.ts:9]: nit:
From what i can tell the long em-dash is put in after the param in JS docs without explicitly adding this dash (if you remove it and mouse over `getNamespaceOfAccountOwner` it still shows an em-dash), should we just omit the dash or have it in all jsdocs? This isn't a very important question, as the behavior is the same, but for consistency we should probably choose one over the other

File: packages/hash/api/src/model/ontology/entity-type.model.ts [hunk 14]
```diff
@@ -285,19 +267,24 @@ export default class {
...
       schema: Omit<EntityType, "$id">;
     },
   ): Promise<EntityTypeModel> {
-    const { accountId, schema } = params;
+    const { schema } = params;
     const updateArguments: UpdateEntityTypeRequest = {
-      accountId,
+      /**
+       * @todo: let caller update who owns the type, or create new method dedicated to changing the owner of the type
+       * @see https://app.asana.com/0/1202805690238892/1203063463721791/f
```
Comment (thehabbos007) [inline packages/hash/api/src/model/ontology/entity-type.model.ts:282]: As before with the `todo`s, this seems to be the same as the todo underneath in the comment. Probably the case for all ontology type `update` methods

File: packages/hash/integration/src/tests/model/knowledge/page.model.test.ts [hunk 15]
```diff
@@ -39,10 +39,10 @@ describe("Page model class", () => {
 
   const createBlock = async () =>
     await BlockModel.createBlock(graphApi, {
-      accountId: testUser.accountId,
+      ownedById: testUser.ownedById,
```
Comment (thehabbos007) [inline packages/hash/integration/src/tests/model/knowledge/page.model.test.ts:42]: Should this be
```suggestion
      ownedById: testUser.entityId,
```
to not get the workspace account?

**Review by thehabbos007 — CHANGES_REQUESTED** [review]
Thanks for doing this, Ben, it's a great clean-up! I have some change requests for things that could lead to problems :)

File: packages/hash/api/src/model/ontology/data-type.model.ts (hunk 11 above)
Comment (benwerner01) [reply packages/hash/api/src/model/ontology/data-type.model.ts]: Thanks for spotting, I intended on linking [this](https://app.asana.com/0/1202805690238892/1203063463721793/f). Must've made a mistake with my clipboard, have addressed in https://github.com/hashintel/hash/pull/1129/commits/5b6dd0e31c8095aab8984efaaa830dd733e1d081

File: packages/hash/api/src/model/ontology/entity-type.model.ts (hunk 14 above)
Comment (benwerner01) [reply packages/hash/api/src/model/ontology/entity-type.model.ts]: Yea, thanks for spotting - have fixed the links in https://github.com/hashintel/hash/pull/1129/commits/5b6dd0e31c8095aab8984efaaa830dd733e1d081

File: packages/hash/api/src/model/ontology/data-type.model.ts (hunk 12 above)
Comment (benwerner01) [reply packages/hash/api/src/model/ontology/data-type.model.ts:172]: It's a task for the Graph API, yes - should I not leave todos where we're waiting for tasks to be completed for the Graph API?

File: packages/hash/api/src/model/ontology/util.ts (hunk 13 above)
Comment (benwerner01) [reply packages/hash/api/src/model/ontology/util.ts:9]: Ideally I think we should use the dashes, because that's what JSDoc does in [their examples](https://jsdoc.app/howto-es2015-classes.html). I also prefer this because it makes the documentation easier to read in places like the GitHub UI why there's no JSDoc syntax highlighting.

Should I create a task to make our JSDocs consistent for the model classes?

File: packages/hash/api/src/model/ontology/data-type.model.ts (hunk 12 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/ontology/data-type.model.ts:172]: Oh sorry, I read the task to be about the workspace API and not the graph API, my bad. I was implying we should change workspace to graph API in the task, but it is about that. I misread :sweat_smile:

File: packages/hash/api/src/model/ontology/util.ts (hunk 13 above)
Comment (thehabbos007) [reply packages/hash/api/src/model/ontology/util.ts:9]: I don't feel strongly about this, you can create a task but probably very low priority to get done, and I assume we have the inconsistency across the codebase

**Review by thehabbos007 — APPROVED** [review]
Thanks Ben! Looks good to me. I think any additions should be follow-ups as this PR is quite big

File: packages/hash/api/src/model/ontology/util.ts (hunk 13 above)
Comment (benwerner01) [reply packages/hash/api/src/model/ontology/util.ts:9]: Yeah, maybe let's try to consistently use the dash with any new JSDoc param definitions we introduce, but otherwise not worry about it too much until we get around to the task. Have created it [here](https://app.asana.com/0/1202805690238892/1203063463721795/f)
